File: dallosh/servers/dallosh_bot/src/lib/sodular/examples/socket_usage.py
"""
Socket usage examples for Sodular client
Exact Python equivalent of examples/socket-usage.ts
"""

import asyncio
from src.lib.sodular import SodularClient


async def socket_usage_example():
    """Example of using sockets with Sodular client"""
    
    # Initialize client
    client = SodularClient({
        'baseUrl': 'http://localhost:5001/api/v1',
        'ai': {'baseUrl': 'http://localhost:4200/api/v1'}
    })
    
    # Connect to server
    result = await client.connect()
    
    if not result.isReady:
        print(f"Failed to connect: {result.error}")
        return
    
    sodular_client = result.client
    
    # Set database context
    sodular_client.use('my_database_id')
    
    # Set authentication token
    sodular_client.setToken('your_access_token')
    
    # Join a channel for a specific table
    sodular_client.joinChannel('my_database_id', 'users_table')
    
    # Listen to events
    def on_user_created(data):
        print(f"User created: {data}")
    
    def on_user_updated(data):
        print(f"User updated: {data}")
    
    def on_user_deleted(data):
        print(f"User deleted: {data}")
    
    # Register event listeners
    sodular_client.on('created', on_user_created)
    sodular_client.on('updated', on_user_updated)
    sodular_client.on('deleted', on_user_deleted)
    
    # Use ref API for table-specific operations
    ref_api = sodular_client.ref.fromTable('users_table')
    
    # Listen to table events
    def on_ref_created(data):
        print(f"Reference created: {data}")
    
    def on_ref_updated(data):
        print(f"Reference updated: {data}")
    
    ref_api.on('created', on_ref_created)
    ref_api.on('updated', on_ref_updated)
    
    # Keep the connection alive
    try:
        await asyncio.sleep(60)  # Keep alive for 60 seconds
    except KeyboardInterrupt:
        print("Stopping socket example...")
    
    # Clean up
    sodular_client.leaveChannel('my_database_id', 'users_table')
    sodular_client.off('created')
    sodular_client.off('updated')
    sodular_client.off('deleted')
    
    # Close client
    await sodular_client.close()


if __name__ == "__main__":
    asyncio.run(socket_usage_example())

File: dallosh/clients/dallosh_web/src/lib/sodular/examples/socket-usage.ts
/**
 * Example: Using Socket Events with RefAPI
 * This demonstrates how to listen to real-time events when refs are created, updated, or deleted
 */

import { SodularClient } from '../index';

async function exampleSocketUsage() {
  // Initialize the client
  const client = SodularClient({
    baseUrl: 'http://localhost:5001/api/v1'
  });

  // Connect to the server
  const { isReady, error, client: sodularClient } = await client.connect();
  
  if (!isReady) {
    console.error('Failed to connect:', error);
    return;
  }

  // Set database context
  sodularClient.use('your-database-id');

  // set token 
  sodularClient.setToken('your-token');

  // Listen to events for a specific table
  const tableId = 'your-table-id';
  
  // Listen to ref creation events
  sodularClient.ref.from(tableId).on('created', (data) => {
    console.log('New ref created:', data);
  });

  // Listen to ref update events (replaced)
  sodularClient.ref.from(tableId).on('replaced', (data) => {
    console.log('Ref replaced:', data);
  });

  // Listen to ref update events (patched)
  sodularClient.ref.from(tableId).on('patched', (data) => {
    console.log('Ref patched:', data);
  });

  // Listen to ref deletion events
  sodularClient.ref.from(tableId).on('deleted', (data) => {
    console.log('Ref deleted:', data);
  });

  // Alternative: Use the client directly for global events
  sodularClient.on('created', (data) => {
    console.log('Global created event:', data);
  });

  // Join a specific channel manually
  sodularClient.joinChannel('your-database-id', 'your-table-id');

  // Later, you can stop listening to specific events
  sodularClient.ref.from(tableId).off('created');

  // Or stop listening to all events for a table
  sodularClient.ref.from(tableId).off();

  // Leave a channel manually
  sodularClient.leaveChannel('your-database-id', 'your-table-id');
}

// Example with error handling
async function exampleWithErrorHandling() {
  try {
    const client = SodularClient({
      baseUrl: 'http://localhost:5001/api/v1'
    });

    const { isReady, error, client: sodularClient } = await client.connect();
    
    if (!isReady) {
      throw new Error(`Connection failed: ${error}`);
    }

    // Set up event listeners
    const tableId = 'users-table';
    
    sodularClient.ref.from(tableId)
      .on('created', (data) => {
        console.log('User created:', data);
        // Handle new user creation
      })
      .on('updated', (data) => {
        console.log('User updated:', data);
        // Handle user updates
      })
      .on('deleted', (data) => {
        console.log('User deleted:', data);
        // Handle user deletion
      });

    console.log('Listening to user table events...');
    
  } catch (error) {
    console.error('Error in socket example:', error);
  }
}

export { exampleSocketUsage, exampleWithErrorHandling };

